### PR TITLE
support run oci image with erofs tarfs

### PR DESCRIPTION
**Pull request by kevinXYin**
Support run OCI image with erofs tarfs

### CODE REVIEW

**Comment by jiangliu** [comment]
We also need to enhance view to support tarfs?

File: pkg/tarfs/tarfs.go [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+// fetch and cache blob digests and diff ids of an oci image
+func (t *Manager) fetchImageDiffID(ctx context.Context, remote *remote.Remote, ref, manifest string) error {
+	// fetch the manifest find config digest and layers digest
+	rc, err := t.getBlobStream(ctx, remote, ref, manifest)
+	if err != nil {
+		return err
+	}
+	var manifestOCI ocispec.Manifest
+	bytes, err := io.ReadAll(rc)
+	rc.Close()
```
Comment (imeoer) [inline pkg/tarfs/tarfs.go:91]:
```
rc, err := t.getBlobStream(ctx, remote, ref, manifest)
if err != nil {
	return err
}
defer rc.Close()
```

File: pkg/tarfs/tarfs.go [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "get remote fetcher")
+	}
+
+	fetcherByDigest, ok := fetcher.(remotes.FetcherByDigest)
+	if !ok {
+		return nil, errors.Errorf("fetcher %T does not implement remotes.FetcherByDigest", fetcher)
+	}
+
+	dgst, err := digest.Parse(layerDigest)
```
Comment (imeoer) [inline pkg/tarfs/tarfs.go:157]: Why not make `layerDigest` as `digest.Digest` type directly.

File: pkg/tarfs/tarfs.go [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+	}
+
+	// fetch the config and find diff ids
+	rc, err = t.getBlobStream(ctx, remote, ref, string(manifestOCI.Config.Digest))
+	if err != nil {
+		return errors.Wrap(err, "fetch referrers")
+	}
+	var config ocispec.Image
+	bytes, err = io.ReadAll(rc)
+	rc.Close()
```
Comment (imeoer) [inline pkg/tarfs/tarfs.go:109]: Ditto.

File: pkg/tarfs/tarfs.go [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+	if err := json.Unmarshal(bytes, &config); err != nil {
+		return errors.Wrap(err, "unmarshal config")
+	}
+	if len(config.RootFS.DiffIDs) != len(manifestOCI.Layers) {
+		return errors.Errorf("number of diff ids unmatch manifest layers")
+	}
+
+	// cache OCI blob digest & diff id
+	for i := range manifestOCI.Layers {
+		t.diffIDCache.Add(string(manifestOCI.Layers[i].Digest), string(config.RootFS.DiffIDs[i]))
```
Comment (imeoer) [inline pkg/tarfs/tarfs.go:122]: `lru.Cache` is not safe for concurrent access.

File: pkg/tarfs/tarfs.go [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+
+	tarFile, err := os.Create(blob)
+	if err != nil {
+		return false, err
+	}
+	defer tarFile.Close()
+
+	fifoName := filepath.Join(storagePath, "layer_"+snapshotID+"_"+"tar.fifo")
+	err = syscall.Mkfifo(fifoName, 0644)
+	if err != nil {
```
Comment (imeoer) [inline pkg/tarfs/tarfs.go:179]: `if err := syscall.Mkfifo(fifoName, 0644); err != nil {`

File: pkg/tarfs/tarfs.go [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+		return false, err
+	}
+	defer os.Remove(fifoName)
+
+	go func() {
+		fifoFile, err := os.OpenFile(fifoName, os.O_WRONLY, os.ModeNamedPipe)
+		if err != nil {
+			return
```
Comment (imeoer) [inline pkg/tarfs/tarfs.go:187]: At least give a warning message for the error.

File: pkg/tarfs/tarfs.go [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+		"--blob-dir", storagePath,
+		fifoName,
+	}
+	cmd := exec.Command(t.nydusImagePath, options...)
+	var errb, outb bytes.Buffer
+	cmd.Stderr = &errb
+	cmd.Stdout = &outb
+	log.L.Debugf("nydus image command %v", options)
+	err = cmd.Run()
+	if err != nil {
```
Comment (imeoer) [inline pkg/tarfs/tarfs.go:209]: The stderr should be redirected to the logger if any error happens.

File: pkg/tarfs/tarfs.go [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+	return loop, err
+}
+
+func (t *Manager) PrepareTarfsLayer(snapshotID, ref, manifest, layerDigest, storagePath string) error {
+	t.mutex.Lock()
+	if _, ok := t.snapshotMap[snapshotID]; ok {
+		t.mutex.Unlock()
+		return errors.Errorf("snapshot %s already prapared", snapshotID)
+	}
+	wg := &sync.WaitGroup{}
```
Comment (imeoer) [inline pkg/tarfs/tarfs.go:297]: What is the meaning of using waitgroup here?

**Comment by kevinXYin** [comment]
@imeoer  Thanks for the detailed review, I will fix these issues and submit an new pr based on https://github.com/containerd/nydus-snapshotter/pull/472 after it merged,  which may be some significant changes.

**Comment by kevinXYin** [comment]
submit a new pr for tarfs https://github.com/containerd/nydus-snapshotter/pull/497 , close this one

**Pull request closed without merging**